src/Provider/AbstractProvider.php

```diff
@@ -387,32 +387,62 @@ protected function verifyGrant($grant)
     }
 
     /**
-     * Returns a prepared request for requesting an access token.
+     * Returns the full URL to use when requesting an access token.
      *
-     * @param array $params Query string parameters
+     * @param array $params Query parameters
      */
-    protected function getAccessTokenRequest(array $params)
+    protected function getAccessTokenUrl(array $params)
     {
         $url = $this->getBaseAccessTokenUrl();
-        $query = $this->getAccessTokenQuery($params);
-        $method = strtoupper($this->getAccessTokenMethod());
 
+        if ($this->getAccessTokenMethod() === self::METHOD_GET) {
+            $query = $this->getAccessTokenQuery($params);
+            return $this->appendQuery($url, $query);
+        }
+
+        return $url;
+    }
+
+    /**
+     * Returns the request body for requesting an access token.
+     *
+     * @param  array $params
+     * @return string
+     */
+    protected function getAccessTokenBody(array $params)
+    {
+        return http_build_query($params);
+    }
+
+    /**
+     * Builds request options used for requesting an access token.
+     *
+     * @param  array $params
+     * @return array
+     */
+    protected function getAccessTokenOptions(array $params)
+    {
         $options = [];
 
-        switch ($method) {
-            case 'GET':
-                $url = $this->appendQuery($url, $query);
-                break;
-            case 'POST':
-                $options['body'] = $query;
-                break;
-            default:
-                throw new InvalidArgumentException(
-                    "Unsupported access token request method: '$method'"
-                );
+        if ($this->getAccessTokenMethod() === self::METHOD_POST) {
+            $options['body'] = $this->getAccessTokenBody($params);
         }
```

Comment:
I still feel that we are losing something by separating GET/POST request building. We're definitely losing method validation, which should be enforced to meet the OAuth 2 RFC.

I guess I'd like to see a much smaller change to allow for the modifications you want, maybe just a one line change?

```php
$options = $this->getAccessTokenOptions($options, $params);
```

Which would be called just before `getRequest` to allow the provider to modify the options without overloading anything else?